cmd/snap-confine/mount-support-nvidia.c

```diff
@@ -438,6 +441,15 @@ static int sc_mount_nvidia_is_driver_in_dir(const char *dir)
 			 "%s/" LIBNVIDIA_GLCORE_SO_PATTERN, dir,
```

Comment:
Then use a single sprintf and check here:
```
// Extra sanity check
if (driver.glcore_pattern == NULL) {
	die("libnvidia-glcore pattern is unset");
}

// Probe if a well known library is found in directory dir
sc_must_snprintf(driver_path, sizeof driver_path,
			 "%s/" driver.glcore_pattern, dir,
			 driver.major_version, driver.minor_version);
```

Reply:
ok, will update the PR

Reply:
cant concat `%s/` and `driver.glcore_pattern` since it's not a constant

Reply:
Yeah, forgot it's actually a format string. I'd go with this:
```
diff --git a/cmd/snap-confine/mount-support-nvidia.c b/cmd/snap-confine/mount-support-nvidia.c
index cbc574afa8..e3362084b8 100644
--- a/cmd/snap-confine/mount-support-nvidia.c
+++ b/cmd/snap-confine/mount-support-nvidia.c
@@ -122,8 +122,7 @@ static const size_t nvidia_globs_len =
     sizeof nvidia_globs / sizeof *nvidia_globs;
 
 #define LIBNVIDIA_GLCORE_SO_PATTERN "libnvidia-glcore.so.%d.%d"
-//the driver version can be `440.33.01` so the pattern should end with `%02d`
-#define LIBNVIDIA_GLCORE_SO_PATTERN_MICRO "libnvidia-glcore.so.%d.%d.%02d"
+#define LIBNVIDIA_GLCORE_SO_PATTERN_MICRO "libnvidia-glcore.so.%d.%d.%d"
 
 #endif                         // defined(NVIDIA_BIARCH) || defined(NVIDIA_MULTIARCH)
 
@@ -351,7 +350,7 @@ struct sc_nvidia_driver {
        int major_version;
        int minor_version;
        int micro_version;
-       char *glcore_pattern;
+       bool has_micro;
 };
 
 static void sc_probe_nvidia_driver(struct sc_nvidia_driver *driver)
@@ -375,18 +374,17 @@ static void sc_probe_nvidia_driver(struct sc_nvidia_driver *driver)
        int res = fscanf(file, "%d.%d.%d", &driver->major_version, 
                                &driver->minor_version, 
                                &driver->micro_version);
-       switch (res) {
-       case 3:
-               driver->glcore_pattern = LIBNVIDIA_GLCORE_SO_PATTERN_MICRO;
-               break;
-       case 2:
-               driver->glcore_pattern = LIBNVIDIA_GLCORE_SO_PATTERN;
-               break;
-       default:
+       if (res < 2) {
                die("cannot parse nvidia driver version string");
        }
-       debug("parsed nvidia driver version: %d.%d.%d", driver->major_version,
-             driver->minor_version, driver->micro_version);
+       if (res == 3) {
+               driver->has_micro = true;
+               debug("parsed nvidia driver version: %d.%d.%d", driver->major_version,
+                         driver->minor_version, driver->micro_version);
+       } else {
+               debug("parsed nvidia driver version: %d.%d", driver->major_version,
+                         driver->minor_version);
+       }
 }
 
 static void sc_mkdir_and_mount_and_bind(const char *rootfs_dir,
@@ -447,26 +445,18 @@ static int sc_mount_nvidia_is_driver_in_dir(const char *dir)
        if (driver.major_version == 0) {
                return 0;
        }
-       // Extra sanity check
-       if (driver.glcore_pattern == NULL) {
-               die("libnvidia-glcore pattern is unset");
-       }
 
        // Probe if a well known library is found in directory dir
-       // cant concat `%s/` and `driver.glcore_pattern` since it's not a constant
-       if (sc_streq(driver.glcore_pattern, LIBNVIDIA_GLCORE_SO_PATTERN)) {
+       if (!driver.has_micro) {
                sc_must_snprintf(driver_path, sizeof driver_path,
                                "%s/" LIBNVIDIA_GLCORE_SO_PATTERN, dir,
                                driver.major_version, driver.minor_version);
-       } else if (sc_streq(driver.glcore_pattern, LIBNVIDIA_GLCORE_SO_PATTERN_MICRO)) {
+       } else {
                sc_must_snprintf(driver_path, sizeof driver_path,
                                "%s/" LIBNVIDIA_GLCORE_SO_PATTERN_MICRO, dir,
                                driver.major_version, driver.minor_version, 
                                driver.micro_version);
-       } else {
-               die("unhandled libnvidia-glcore pattern");
        }
-       
 
        if (access(driver_path, F_OK) == 0) {
                debug("nvidia library detected at path %s", driver_path);
```

Reply:
i was actually thinking of removing `struct sc_nvidia_driver` altogether and just keep the driver string as it is stored in `/sys/module/nvidia/version`
pros: no parsing, no workarounds for version numbers not following semver altogether e.g. `1.0.42a1`
cons: ideas ?

Reply:
Come to think about that, maybe that would be the best course of action. There's no guarantee the next release is not going to have `123.01.0001` as the version number. Capturing this in a proper format string for the globs feels a bit error prone.

As long as we do some verification in `sc_probe_nvidia_driver` of the format it we could drop sc_nvidia_driver. I would suggest doing something like this:

```
#define SC_NVIDIA_VERSION_MAX_LEN 20

static bool sc_probe_nvidia_driver_version(char *version, size_t version_len) {
	if () { // no driver file 
		return false;
	}
	// read the line, eg. fgets(version, version_len, file);
	// validate the format,. eg. sscanf(version, "%d.%d.%d")
	return true;
}

#define LIBNVIDIA_GLCORE_SO_PATTERN "libnvidia-glcore.so."

static int sc_mount_nvidia_is_driver_in_dir(const char *dir) {
	...
	const char driver_version[SC_NVIDIA_VERSION_MAX_LEN+1] = { 0 };
	iif (!sc_probe_nvidia_driver_version(driver_version, sizeof driver_version)) {
		return 0;
	}
	sc_must_snprintf(driver_path, sizeof driver_path,
			 "%s/" LIBNVIDIA_GLCORE_SO_PATTERN "%s", dir,
			 driver_version);
	....
}
```

Reply:
while rewriting the code to remove `struct sc_nvidia_driver` i realized that we also need to link `nvidia-$MAJOR/xorg` to `gl`

Reply:
`	sc_must_snprintf(src, sizeof src, "%s-%d", src_dir, driver.major_version);`